File: julia/src/callables.jl
# The callable protocol of section 10, and the `affine` reference
# callable of section 27.
#
# A callable is exactly four things and nothing more: call, to_dict,
# from_dict, and an optional one-line repr.  Everything else it knows
# lives inside its own dictionary and is its own business.

"""
    Prediction(mean; uncertainty = nothing, level = nothing, method = nothing)

What a callable returns for one output (section 10).  `mean` is the
point prediction, held the way this package holds every array: the
Julia axes reversed from the file's, so a scalar slot's mean is a
vector over rows and an array slot's is (component, node, row).
`uncertainty` is optional: a band (section 9), the half-width of the
interval around the mean at coverage `level`, of the same shape, made
as `method` says.  The three go together, and a band missing its level
or its method is refused here rather than in a file.
"""
struct Prediction
    mean::Array{Float64}
    uncertainty::Union{Nothing,Array{Float64}}
    level::Union{Nothing,Float64}
    method::Union{Nothing,String}

    function Prediction(mean; uncertainty = nothing, level = nothing,
                        method = nothing)
        m = convert(Array{Float64}, mean)
        if uncertainty === nothing
            (level === nothing && method === nothing) ||
                throw(MestraError(nothing,
                    "level and method go with an uncertainty; a " *
                    "prediction without one has neither (section 10)"))
            return new(m, nothing, nothing, nothing)
        end
        u = convert(Array{Float64}, uncertainty)
        size(u) == size(m) || throw(MestraError(nothing,
            "a band has the shape of its mean; the mean is $(size(m)) " *
            "and the band $(size(u))"))
        (level isa Real && 0 < level < 1) || throw(MestraError(nothing,
            "a band states the coverage it claims as level in (0, 1); a " *
            "1.96-sigma Gaussian band is 0.95"))
        (method isa AbstractString && !isempty(method)) ||
            throw(MestraError(nothing,
                "a band says how it was made; give method one sentence"))
        any(x -> x < 0, u) && throw(MestraError(nothing,
            "a band is a half-width and is never negative"))
        return new(m, u, Float64(level), String(method))
    end
end

"""True when the record carries a band."""
has_uncertainty(p::Prediction) = p.uncertainty !== nothing

"""
    Callable

Keys in, predictions out.  A type conforming to the protocol defines

    (c::MyCallable)(keys)        -> Dict{String,Prediction}
    Mestra.to_dict(c)            -> Dict{String,Any}
    Mestra.from_dict(::Type{MyCallable}, d) -> MyCallable

and registers itself with `register_callable!("my type", MyCallable)`.
`Base.show(io, c)` is optional and is the one-line `repr` the file may
carry.

`call` returns one `Prediction` per output the callable serves, named
by the value of the slot's `output` attribute, whose mean is shaped as
the slot would be stored, (row, node | cell, component) for an array
and (row) for a scalar, and which carries a band when the model has
one.  How the band was computed is the model's business; what it
claims is on the record.
"""
abstract type Callable end

const CALLABLE_REGISTRY = Dict{String,Type}()

"""
    register_callable!(type, T)

Make `from_dict` dispatch on the `type` string a file carries.
"""
function register_callable!(type::AbstractString, T::Type)
    CALLABLE_REGISTRY[String(type)] = T
    return T
end

"""The Julia type registered for a `type` string, or nothing."""
callable_type(type::AbstractString) =
    get(CALLABLE_REGISTRY, String(type), nothing)

"""
    to_dict(c::Callable) -> Dict{String,Any}

The nested dictionary that fully represents the callable.  `type` and
`repr` are the container's attributes and are never in it (section 25).
"""
function to_dict end

"""
    from_dict(T, dict) -> Callable

The inverse of `to_dict`, dispatched on the `type` string through the
registry.
"""
function from_dict end

"""
    build_callable(ref::CallableRef) -> Callable

Turn what the file carries into a callable, if this reader owns the
type.  A reader that does not own it may still copy the dictionary and
must not interpret it (section 12), which is what `ref.dict` is for.
"""
function build_callable(ref::CallableRef)
    ref.type === nothing && throw(MestraError("E15",
        "callable $(ref.id) has no `type` attribute"))
    T = callable_type(ref.type)
    T === nothing && throw(MestraError(nothing,
        "no callable registered for type $(ref.type); the dictionary " *
        "is opaque to this reader and must not be interpreted"))
    return from_dict(T, ref.dict)
end

# -------------------------------------------------------- keys tables

"""
    KeysTable

The keys table of section 26 is, in Julia, a `Dict{String,Vector{Float64}}`
of key name to column, with every column the same length and the row
order the evaluation order: output row i is the result for table row i.
A key with role `id` may hold a `Vector{String}` instead, so the
general element type is `AbstractVector`.

`normalise_keys` also accepts a `NamedTuple` of columns and a
`(matrix, names)` pair, where the matrix is (rows, keys) and the names
are in the file's key order, so that a caller may pass whichever is to
hand.
"""
const KeysTable = AbstractDict{String,<:AbstractVector}

normalise_keys(t::AbstractDict) =
    Dict{String,AbstractVector}(String(k) => v for (k, v) in t)

normalise_keys(t::NamedTuple) =
    Dict{String,AbstractVector}(String(k) => v for (k, v) in pairs(t))

function normalise_keys(t::Tuple{AbstractMatrix,Any})
    m, names = t
    size(m, 2) == length(names) || throw(MestraError(nothing,
        "a keys matrix of $(size(m, 2)) columns with $(length(names)) names"))
    return Dict{String,AbstractVector}(String(names[j]) => collect(m[:, j])
                                       for j in 1:length(names))
end

function keys_table_rows(t::AbstractDict)
    isempty(t) && return 0
    lengths = unique(length.(collect(Base.values(t))))
    length(lengths) == 1 || throw(MestraError(nothing,
        "the columns of a keys table must all be the same length"))
    return lengths[1]
end

# ------------------------------------------------------------- affine

"""One output of an `affine` callable: y = A x + b, reshaped to
`shape` in C order (section 27), and, when the output carries a band,
a constant `uncertainty` the same in every row with its `level` and
`method`: all three or none."""
struct AffineOutput
    A::Matrix{Float64}      # (n_out_flat, n_keys)
    b::Vector{Float64}      # (n_out_flat,)
    shape::Vector{Int64}    # the slot's dimensions after the row one
    uncertainty::Union{Nothing,Vector{Float64}}   # (n_out_flat,)
    level::Union{Nothing,Float64}
    method::Union{Nothing,String}
end

AffineOutput(A, b, shape) = AffineOutput(A, b, shape, nothing, nothing, nothing)

has_band(o::AffineOutput) = o.uncertainty !== nothing

"""Section 27: a band is all three of uncertainty, level and method, the
uncertainty is one value per output element, and the level is a
coverage."""
function check_band(name, o::AffineOutput)
    parts = (o.uncertainty !== nothing) + (o.level !== nothing) +
            (o.method !== nothing)
    parts == 0 && return o
    parts == 3 || throw(MestraError(nothing,
        "a band on the affine output `$(name)` is all three of " *
        "uncertainty, level and method"))
    length(o.uncertainty) == n_out_flat(o) || throw(MestraError(nothing,
        "the affine output `$(name)` has an uncertainty of " *
        "$(length(o.uncertainty)) values where its shape needs " *
        "$(n_out_flat(o))"))
    0 < o.level < 1 || throw(MestraError(nothing,
        "the affine output `$(name)` has a level that is not a coverage " *
        "in (0, 1)"))
    isempty(o.method) && throw(MestraError(nothing,
        "the affine output `$(name)` has a method that is not one sentence"))
    return o
end

"""
    Affine(keys, outputs)

The one callable type this package defines, so that the protocol, the
codec and evaluation can be conformance tested with no proprietary
model.  `keys` is the declared key order and `outputs` maps the value
of a slot's `output` attribute to its A, b and shape.
"""
struct Affine <: Callable
    keys::Vector{String}
    outputs::Dict{String,AffineOutput}

    function Affine(keys, outputs)
        for (name, o) in outputs
            check_band(name, o)
        end
        return new(keys, outputs)
    end
end

Base.show(io::IO, c::Affine) =
    print(io, "affine(", join(c.keys, ", "), " -> ",
          join(sort(collect(keys(c.outputs)), by = codeunits), ", "), ")")

n_out_flat(o::AffineOutput) = isempty(o.shape) ? 1 : Int(prod(o.shape))

"""
    (c::Affine)(table) -> Dict{String,Prediction}

Evaluate every output on a keys table.  The dot product is accumulated
over the keys in the declared key order and b is added last, with no
fused multiply-add (section 27).  The loop below is
written out for that reason: it uses Float64 throughout, it never
calls `muladd`, and it carries no `@simd` or `@fastmath`, either of
which would let the compiler reassociate or contract it.  Julia's
floating-point semantics forbid the compiler from fusing `acc + a * x`
into an FMA on its own, so the order in the source is the order that
runs.
"""
function (c::Affine)(table)
    t = normalise_keys(table)
    nrows = keys_table_rows(t)
    x = Matrix{Float64}(undef, nrows, length(c.keys))
    for (j, k) in pairs(c.keys)
        haskey(t, k) || throw(MestraError(nothing,
            "the keys table has no column `$(k)`, which this callable " *
            "declares"))
        col = t[k]
        length(col) == nrows || throw(MestraError(nothing,
            "column `$(k)` has $(length(col)) rows, not $(nrows)"))
        for i in 1:nrows
            x[i, j] = Float64(col[i])
        end
    end
    out = Dict{String,Prediction}()
    for (name, o) in c.outputs
        size(o.A, 2) == length(c.keys) || throw(MestraError(nothing,
            "output `$(name)` has an A of $(size(o.A, 2)) columns for " *
            "$(length(c.keys)) keys"))
        nf = n_out_flat(o)
        size(o.A, 1) == nf && length(o.b) == nf || throw(MestraError(nothing,
            "output `$(name)` has A and b that do not match its shape"))
        y = Matrix{Float64}(undef, nrows, nf)
        for r in 1:nrows
            for i in 1:nf
                acc = 0.0
                for k in 1:length(c.keys)
                    acc = acc + o.A[i, k] * x[r, k]
                end
                acc = acc + o.b[i]
                y[r, i] = acc
            end
        end
        mean = reshape_output(y, o.shape)
        if has_band(o)
            # The band is a constant, repeated for each row, so that
            # evaluation stays exact (section 27).
            u = repeat(reshape(o.uncertainty, 1, nf), nrows, 1)
            out[name] = Prediction(mean; uncertainty = reshape_output(u, o.shape),
                                   level = o.level, method = o.method)
        else
            out[name] = Prediction(mean)
        end
    end
    return out
end

"""Give a (rows, n_out_flat) block the slot's stored shape.  The
result is held the way this package holds every array: the Julia axes
reversed from the file's, so its linear memory is the file's order."""
function reshape_output(y::Matrix{Float64}, shape::Vector{Int64})
    nrows = size(y, 1)
    cdims = vcat(nrows, Int.(shape))
    # y is (row, flat) with flat contiguous in C order within a row,
    # which is the file's order, so the transpose is already it.
    flat = vec(permutedims(y, (2, 1)))
    return reshape(flat, reverse(cdims)...)
end

function output_dict(o::AffineOutput)
    d = Dict{String,Any}("A" => copy(o.A), "b" => copy(o.b),
                         "shape" => copy(o.shape))
    if has_band(o)
        # Section 25: the band is a dataset, and the level and the
        # method are attributes on the entry.
        d["uncertainty"] = copy(o.uncertainty)
        d["level"] = o.level
        d["method"] = o.method
    end
    return d
end

to_dict(c::Affine) = Dict{String,Any}(
    "keys" => copy(c.keys),
    "outputs" => Dict{String,Any}(name => output_dict(o)
                                  for (name, o) in c.outputs))

const AFFINE_OUTPUT_KEYS = Set(["A", "b", "shape", "uncertainty", "level",
                                "method"])

function from_dict(::Type{Affine}, d::AbstractDict)
    extra = setdiff(Set(keys(d)), Set(["keys", "outputs"]))
    isempty(extra) || throw(MestraError(nothing,
        "an `affine` dictionary with keys section 27 does not define: " *
        join(sort(collect(extra)), ", ")))
    haskey(d, "keys") && haskey(d, "outputs") || throw(MestraError(nothing,
        "an `affine` dictionary needs `keys` and `outputs`"))
    ks = String[String(k) for k in d["keys"]]
    outs = Dict{String,AffineOutput}()
    for (name, entry) in d["outputs"]
        e = setdiff(Set(keys(entry)), AFFINE_OUTPUT_KEYS)
        isempty(e) || throw(MestraError(nothing,
            "an `affine` output with keys section 27 does not define: " *
            join(sort(collect(e)), ", ")))
        A = Float64.(entry["A"])
        ndims(A) == 2 || throw(MestraError(nothing,
            "an `affine` A must be two-dimensional"))
        outs[String(name)] = AffineOutput(
            A, Float64.(vec(entry["b"])), Int64.(vec(entry["shape"])),
            haskey(entry, "uncertainty") ?
                Float64.(vec(entry["uncertainty"])) : nothing,
            haskey(entry, "level") ? Float64(entry["level"]) : nothing,
            haskey(entry, "method") ? String(entry["method"]) : nothing)
    end
    return Affine(ks, outs)
end

"""
    affine(keys, outputs) -> Affine

Build an `affine` callable from plain arrays, for example

    affine(["mach", "alpha"],
           Dict("cl" => (A = [2.0 0.1], b = [0.05], shape = Int64[])))

An output with a band gives `uncertainty`, `level` and `method` too.
"""
function affine(keys, outputs::AbstractDict)
    outs = Dict{String,AffineOutput}()
    for (name, o) in outputs
        A = Float64.(o.A)
        outs[String(name)] = AffineOutput(
            ndims(A) == 2 ? A : reshape(A, 1, length(A)),
            Float64.(vec(o.b)), Int64.(vec(o.shape)),
            hasproperty(o, :uncertainty) ? Float64.(vec(o.uncertainty)) :
                nothing,
            hasproperty(o, :level) ? Float64(o.level) : nothing,
            hasproperty(o, :method) ? String(o.method) : nothing)
    end
    return Affine(String[String(k) for k in keys], outs)
end

File: python/mestra/callables.py
"""The callable protocol, its registry, and the `affine` reference.

A callable is four things and nothing more (SPEC.md section 10):
`__call__` takes a keys table and returns one `Prediction` per output
it serves, `to_dict` represents it as a nested dictionary, `from_dict`
is the inverse dispatched on a `type` string, and `__repr__` is an
optional one-line description. Everything else it knows is inside its
own dictionary and is its own business.

A `Prediction` is a mean and, when the model has one, a band: the
half-width `uncertainty` of the interval around the mean at coverage
`level`, made as `method` says. How the band was computed is the
model's business; what it claims is on the record.

Adding a type is a subclass and one registration call:

    class Lookup(mestra.Callable):
        type = "lookup"
        def __call__(self, keys): ...
        def to_dict(self): ...
        @classmethod
        def from_dict(cls, d): ...

    mestra.register_callable(Lookup)
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any, ClassVar

import numpy as np

from .errors import MestraError

__all__ = [
    "Prediction",
    "Callable",
    "Affine",
    "OpaqueCallable",
    "register_callable",
    "callable_types",
    "callable_from_dict",
    "keys_table",
    "table_length",
]


# ------------------------------------------------------- the keys table

def keys_table(table: Any, names: Sequence[str] | None = None
               ) -> dict[str, np.ndarray]:
    """Normalise a keys table to a mapping of name to column.

    Section 26 gives Python two spellings: a mapping from key name to
    a one-dimensional array, every array the same length, or a
    two-dimensional array of shape (rows, keys) together with the key
    names in the file's key order. Both are accepted here.
    """
    if isinstance(table, Mapping):
        out = {str(name): _column(name, values)
               for name, values in table.items()}
    else:
        values = np.asarray(table)
        if values.ndim != 2:
            raise MestraError(
                "", "a keys table is a mapping of name to column, or a "
                "two-dimensional array with the key names (section 26)")
        if names is None or len(names) != values.shape[1]:
            raise MestraError(
                "", "a two-dimensional keys table needs one name per "
                "column, in the file's key order; pass names="
                "(section 26)")
        out = {str(name): values[:, at]
               for at, name in enumerate(names)}
    lengths = {len(column) for column in out.values()}
    if len(lengths) > 1:
        raise MestraError(
            "", "every column of a keys table has the same length; "
            "these have %s"
            % ", ".join(str(n) for n in sorted(lengths)))
    return out


def _column(name: Any, values: Any) -> np.ndarray:
    column = np.asarray(values)
    if column.ndim == 0:
        column = column.reshape(1)
    if column.ndim != 1:
        raise MestraError(
            "", "the column of a keys table has one dimension, row",
            str(name))
    return column


def table_length(table: Mapping[str, np.ndarray]) -> int:
    """The number of rows in a normalised keys table."""
    for column in table.values():
        return int(len(column))
    return 0


# -------------------------------------------------------- the record

@dataclass(frozen=True)
class Prediction:
    """What a callable returns for one output (section 10).

    `mean` is the point prediction, shaped as the slot is stored:
    (row, node | cell, component) for an array, (row) for a scalar.
    `uncertainty` is optional: a band, the half-width of the interval
    around `mean` at coverage `level`, of the same shape, made as
    `method` says. The three go together, and a band that is missing
    its level or its method is refused here rather than in a file.

        Prediction(mean)
        Prediction(mean, band, level=0.95, method="...")
    """

    mean: np.ndarray
    uncertainty: np.ndarray | None = None
    level: float | None = None
    method: str | None = None

    def __post_init__(self) -> None:
        mean = np.asarray(self.mean, dtype="<f8")
        object.__setattr__(self, "mean", mean)
        if self.uncertainty is None:
            if self.level is not None or self.method is not None:
                raise MestraError(
                    "section 10", "level and method go with an "
                    "uncertainty; a prediction without one has neither")
            return
        band = np.asarray(self.uncertainty, dtype="<f8")
        if band.shape != mean.shape:
            raise MestraError(
                "section 10", "a band has the shape of its mean; the "
                "mean is %s and the band %s"
                % (mean.shape, band.shape))
        if self.level is None or not (0.0 < float(self.level) < 1.0):
            raise MestraError(
                "section 10", "a band states the coverage it claims as "
                "level in (0, 1), and this one gives %r" % (self.level,))
        if not self.method:
            raise MestraError(
                "section 10", "a band says how it was made; give method "
                "one sentence")
        if bool((band < 0).any()):
            raise MestraError(
                "section 10", "a band is a half-width and is never "
                "negative")
        object.__setattr__(self, "uncertainty", band)
        object.__setattr__(self, "level", float(self.level))
        object.__setattr__(self, "method", str(self.method))

    @property
    def has_uncertainty(self) -> bool:
        """True when the record carries a band."""
        return self.uncertainty is not None


# ---------------------------------------------------------- the protocol

class Callable(ABC):
    """Keys in, predictions out, plus a dictionary that represents it.

    A subclass sets `type` to the string that names it in a file and
    implements the three methods below.
    """

    #: The public type string, so that a reader knows which tool can
    #: evaluate the callable (section 10).
    type: ClassVar[str] = ""

    @abstractmethod
    def __call__(self, keys: Any) -> dict[str, Prediction]:
        """One prediction per output this callable serves.

        `keys` is a keys table (section 26). The result is a mapping
        from an output name to a `Prediction` whose `mean` is shaped
        as the slot would be stored, (row, node | cell, component)
        for an array and (row) for a scalar, with a band when the
        model has one.
        """

    @abstractmethod
    def to_dict(self) -> dict[str, Any]:
        """A nested dictionary of arrays, numbers and strings that
        fully represents this callable (sections 10 and 17)."""

    @classmethod
    @abstractmethod
    def from_dict(cls, d: Mapping[str, Any]) -> Callable:
        """Rebuild a callable from the dictionary `to_dict` made."""


_REGISTRY: dict[str, type[Callable]] = {}


def register_callable(kind: type[Callable]) -> type[Callable]:
    """Register a callable class under its `type` string.

    Usable as a decorator. A second registration of the same type
    replaces the first, so a tool can override a reference type with
    its own.
    """
    if not getattr(kind, "type", ""):
        raise MestraError(
            "E15", "a callable class must set a type string",
            kind.__name__)
    _REGISTRY[kind.type] = kind
    return kind


def callable_types() -> dict[str, type[Callable]]:
    """The registered callable types, by their type string."""
    return dict(_REGISTRY)


def callable_from_dict(kind: str, d: Mapping[str, Any],
                       repr_line: str | None = None) -> Any:
    """Rebuild a callable of type `kind`, or keep it opaque.

    A reader that does not own the type may still copy the
    dictionary and must not interpret it (section 12), which is what
    `OpaqueCallable` is for.
    """
    known = _REGISTRY.get(kind)
    if known is None:
        return OpaqueCallable(kind, d, repr_line)
    return known.from_dict(d)


class OpaqueCallable:
    """A callable whose type this reader does not own.

    It round-trips the dictionary unchanged and refuses to be
    called.
    """

    def __init__(self, kind: str, d: Mapping[str, Any],
                 repr_line: str | None = None) -> None:
        self.type = kind
        self._dict = dict(d)
        self._repr = repr_line

    def __call__(self, keys: Any) -> dict[str, Prediction]:
        raise MestraError(
            "section 10", "this reader does not own the callable type "
            "%r, so it can copy it but not evaluate it" % self.type)

    def to_dict(self) -> dict[str, Any]:
        """The dictionary as it was read, unchanged."""
        return dict(self._dict)

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> OpaqueCallable:
        return cls("", d)

    def __repr__(self) -> str:
        if self._repr:
            return self._repr
        return "OpaqueCallable(type=%r)" % self.type


# ------------------------------------------------------------- affine

_AFFINE_KEYS = frozenset(["A", "b", "shape"])
_BAND_KEYS = frozenset(["uncertainty", "level", "method"])


@register_callable
class Affine(Callable):
    """y = A x + b per slot: the one callable type this package
    defines (section 27).

    It exists so that the protocol, the codec and evaluation can be
    conformance tested with no proprietary model. It is
    deterministic. An output may carry a constant band, the same in
    every row, given as `uncertainty`, `level` and `method` together.

        m = Affine(["mach", "alpha"],
                   {"cl": {"A": [[2.0, 0.1]], "b": [0.05],
                           "shape": []}})
        m({"mach": [0.5], "alpha": [4.0]})["cl"].mean
    """

    type: ClassVar[str] = "affine"

    def __init__(self, keys: Sequence[str],
                 outputs: Mapping[str, Mapping[str, Any]],
                 repr_line: str | None = None) -> None:
        self.keys = [str(k) for k in keys]
        self.outputs: dict[str, dict[str, np.ndarray]] = {}
        for name, entry in outputs.items():
            extra = set(entry) - _AFFINE_KEYS - _BAND_KEYS
            if extra:
                raise MestraError(
                    "section 27", "an affine output holds A, b and "
                    "shape, and a band as uncertainty, level and "
                    "method, and nothing else; this one also holds %s"
                    % ", ".join(sorted(extra)), str(name))
            band_keys = set(entry) & _BAND_KEYS
            if band_keys and band_keys != _BAND_KEYS:
                raise MestraError(
                    "section 27", "a band on an affine output is all "
                    "three of uncertainty, level and method; this one "
                    "has %s" % ", ".join(sorted(band_keys)), str(name))
            shape = np.asarray(entry["shape"], dtype="<i8").reshape(-1)
            matrix = np.asarray(entry["A"], dtype="<f8")
            offset = np.asarray(entry["b"], dtype="<f8").reshape(-1)
            flat = int(np.prod(shape)) if shape.size else 1
            if matrix.shape != (flat, len(self.keys)):
                raise MestraError(
                    "section 27", "A is (%d, %d) for this output and "
                    "the file gives %s"
                    % (flat, len(self.keys), matrix.shape), str(name))
            if offset.shape != (flat,):
                raise MestraError(
                    "section 27", "b is (%d,) for this output and the "
                    "file gives %s" % (flat, offset.shape), str(name))
            made: dict[str, Any] = {"A": matrix, "b": offset,
                                    "shape": shape}
            if band_keys:
                band = np.asarray(entry["uncertainty"],
                                  dtype="<f8").reshape(-1)
                if band.shape != (flat,):
                    raise MestraError(
                        "section 27", "uncertainty is (%d,) for this "
                        "output and the file gives %s"
                        % (flat, band.shape), str(name))
                level = float(entry["level"])
                if not (0.0 < level < 1.0):
                    raise MestraError(
                        "section 27", "level is a coverage in (0, 1), "
                        "and this output gives %r" % (level,), str(name))
                method = str(entry["method"])
                if not method:
                    raise MestraError(
                        "section 27", "method is one sentence, not the "
                        "empty string", str(name))
                made.update(uncertainty=band, level=level, method=method)
            self.outputs[str(name)] = made
        self._repr = repr_line

    # -- the protocol

    def __call__(self, keys: Any) -> dict[str, Prediction]:
        """Evaluate every output on a keys table (section 27)."""
        table = keys_table(keys)
        missing = [k for k in self.keys if k not in table]
        if missing:
            raise MestraError(
                "", "the keys table has no column %s; this callable "
                "reads %s" % (", ".join(missing),
                              ", ".join(self.keys)))
        rows = table_length(table)
        columns = np.empty((rows, len(self.keys)), dtype="<f8")
        for at, name in enumerate(self.keys):
            columns[:, at] = np.asarray(table[name], dtype="<f8")
        return {name: self._predict(entry, columns)
                for name, entry in self.outputs.items()}

    def to_dict(self) -> dict[str, Any]:
        """The dictionary of section 27, and nothing else."""
        return {
            "keys": list(self.keys),
            "outputs": {name: {key: entry[key]
                               for key in ("A", "b", "shape",
                                           "uncertainty", "level",
                                           "method") if key in entry}
                        for name, entry in self.outputs.items()},
        }

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> Affine:
        extra = set(d) - {"keys", "outputs"}
        if extra:
            raise MestraError(
                "section 27", "an affine dictionary holds keys and "
                "outputs and nothing else; this one also holds %s"
                % ", ".join(sorted(extra)))
        keys = [str(k) for k in d["keys"]]
        return cls(keys, d["outputs"])

    def __repr__(self) -> str:
        if self._repr:
            return self._repr
        return "affine(%s -> %s)" % (", ".join(self.keys),
                                     ", ".join(sorted(self.outputs)))

    # -- the arithmetic, in the order section 27 fixes

    @classmethod
    def _predict(cls, entry: Mapping[str, Any],
                 columns: np.ndarray) -> Prediction:
        """The mean, and the constant band repeated over the rows."""
        mean = cls._evaluate(entry, columns)
        if "uncertainty" not in entry:
            return Prediction(mean)
        band = np.asarray(entry["uncertainty"]).reshape(mean.shape[1:])
        return Prediction(
            mean, np.broadcast_to(band, mean.shape).copy(),
            level=entry["level"], method=entry["method"])

    @staticmethod
    def _evaluate(entry: Mapping[str, Any],
                  columns: np.ndarray) -> np.ndarray:
        """Y = X A' + b, accumulated in the declared key order.

        The dot product is accumulated over the keys in the declared
        order and b is added last, with no fused multiply-add,
        because the corpus compares float64 results bit for bit and
        the other orders differ in the last place.
        """
        matrix = entry["A"]
        rows = columns.shape[0]
        out = np.zeros((rows, matrix.shape[0]), dtype="<f8")
        for at in range(matrix.shape[1]):
            out += columns[:, at:at + 1] * matrix[:, at]
        out += entry["b"]
        shape = tuple(int(n) for n in entry["shape"])
        return out.reshape((rows,) + shape)

    # -- building one

    @classmethod
    def from_arrays(cls, keys: Sequence[str],
                    outputs: Mapping[str, tuple[Any, Any, Sequence[int]]]
                    ) -> Affine:
        """Build from (A, b, shape) triples, one per output."""
        return cls(keys, {name: {"A": a, "b": b, "shape": shape}
                          for name, (a, b, shape) in outputs.items()})
